internal/pkg/cli/local_run.go

```diff
@@ -406,6 +443,141 @@ func (o *localRunOpts) runContainers(ctx context.Context, imageInfoList []clidep
 	return nil
 }
 
+type envVarValue struct {
+	Value  string
+	Secret bool
+}
+
+// getEnvVars uses env overrides passed by flags and environment variables/secrets
+// specified in the Task Definition to return a set of environment varibles for each
+// continer defined in the TaskDefinition. The returned map is a map of container names,
+// each of which contains a mapping of key->envVarValue, which defines if the variable is a secret or not.
+func (o *localRunOpts) getEnvVars(ctx context.Context, taskDef *awsecs.TaskDefinition) (map[string]map[string]envVarValue, error) {
+	envVars := make(map[string]map[string]envVarValue)
+	for _, ctr := range taskDef.ContainerDefinitions {
+		envVars[aws.StringValue(ctr.Name)] = make(map[string]envVarValue)
+	}
+
+	if err := o.fillEnvOverrides(envVars); err != nil {
+		return nil, fmt.Errorf("parse env overrides: %w", err)
+	}
+
+	for _, e := range taskDef.EnvironmentVariables() {
+		// only set if it wasn't overridden
+		if _, ok := envVars[e.Container][e.Name]; !ok {
+			envVars[e.Container][e.Name] = envVarValue{
+				Value: e.Value,
+			}
+		}
+	}
+
+	if err := o.fillSecrets(ctx, envVars, taskDef); err != nil {
+		return nil, fmt.Errorf("get secrets: %w", err)
+	}
+	return envVars, nil
+}
+
+// fillEnvOverrides parses environment variable overrides passed via flag.
+// The expected format of the flag values is KEY=VALUE, with an optional container name
+// in the format of [containerName]:KEY=VALUE. If the container name is omitted,
+// the environment variable override is applied to all containers in the task definition.
+func (o *localRunOpts) fillEnvOverrides(envVars map[string]map[string]envVarValue) error {
+	for k, v := range o.envOverrides {
+		if !strings.Contains(k, ":") {
+			// apply override to all containers
+			for ctr := range envVars {
+				envVars[ctr][k] = envVarValue{
+					Value: v,
+				}
+			}
+			continue
+		}
+
+		// only apply override to the specified container
+		split := strings.SplitN(k, ":", 2)
+		ctr, key := split[0], split[1] // len(split) will always be 2 since we know there is a ":"
+		if _, ok := envVars[ctr]; !ok {
+			return fmt.Errorf("%q targets invalid container", k)
+		}
+		envVars[ctr][key] = envVarValue{
+			Value: v,
+		}
+	}
+
+	return nil
+}
+
+// fillSecrets collects non-overridden secrets from the task definition and
+// makes requests to SSM and Secrets Manager to get their value.
+func (o *localRunOpts) fillSecrets(ctx context.Context, envVars map[string]map[string]envVarValue, taskDef *awsecs.TaskDefinition) error {
+	// figure out which secrets we need to get, set value to ValueFrom
+	unique := make(map[string]string)
+	for _, s := range taskDef.Secrets() {
+		// only set if it wasn't overridden
```

Comment:
so if the env var is defined by variable, we won't try to retrieve the secret? For example, `FOO` is defined in both env var and secret. Shall we error out instead?

Reply:
i verified this behavior with ECS, and you're right! If you define a secret that is already defined by an environment variable, ECS throws this error:
`Create TaskDefinition: The secret name must be unique and not shared with any new or existing environment variables set on the container, such as 'TEST'`

so i added a similar error if we come across an non-overridden env var!